### ISSUE TITLE
你好关于boost fiber报错的问题

### ISSUE THREAD

**Issue by sephirothty** [jxfwinter/jkit#1]
#define BOOST_COROUTINES_NO_DEPRECATION_WARNING
#include <QCoreApplication>
#include <boost/asio.hpp>
#include <boost/asio/spawn.hpp>
#include <boost/bind.hpp>
#include <boost/shared_ptr.hpp>
#include <boost/make_shared.hpp>
#include <string>
#include <ctime>
#include <iostream>
#include <boost/asio/post.hpp>
#include <boost/enable_shared_from_this.hpp>

using namespace boost::asio;
using std::cerr;
using std::endl;

#include <boost/asio/ip/udp.hpp>
#include <boost/fiber/all.hpp>
#include <boost/asio/io_context.hpp>
#include <boost/asio/ip/tcp.hpp>
#include <boost/asio/spawn.hpp>
#include <boost/asio/steady_timer.hpp>
#include <boost/asio/write.hpp>
#include <boost/bind.hpp>
#include <boost/shared_ptr.hpp>
#include <boost/enable_shared_from_this.hpp>
#include <iostream>
#include <thread>
#include <functional>
#include <boost/fiber/future.hpp>
#include <future>
#include <boost/asio/use_fiber_future.hpp>

using boost::asio::ip::tcp;

class session : public boost::enable_shared_from_this<session>
{
public:
  explicit session(boost::asio::io_context& io_context)
    : strand_(io_context),
      socket_(io_context),
      timer_(io_context)
  {
  }

  tcp::socket& socket()
  {
    return socket_;
  }

  void go(boost::asio::io_context& io)
  {

//    boost::asio::spawn(strand_,
//        boost::bind(&session::echo,
//          shared_from_this(),boost::ref(io), _1));
//     std::cout<<"go:"<<std::this_thread::get_id()<<std::endl;
//    boost::asio::spawn(strand_,
//        boost::bind(&session::timeout,
//          shared_from_this(), _1));

      boost::fibers::fiber([&]{

        shared_from_this();
        boost::fibers::future<std::size_t> f;
        boost::system::error_code ec;
        std::future<std::size_t> fut ;
        while(true)
        {
            try
            {
                f = socket_.async_read_some(boost::asio::buffer(data), boost::asio::fibers::use_future([](const boost::system::error_code& ec, std::size_t n){
                                              return ec ? 0 : n;
                                          }));

               socket_.async_write_some(boost::asio::buffer(data), boost::asio::fibers::use_future([](boost::system::error_code ec, size_t n) {
                                           return ec;
                                       }));
            }
            catch (std::system_error& e)
            {
                std::cout << e.what() << "," << typeid(e).name();
                break;
            }

        }
      });

  }

private:
  void do_read()
  {

  }
//  void echo(boost::asio::io_context& io_context,boost::asio::yield_context yield)
//  {
//    try
//    {

//          std::cout<<"echo:"<<std::this_thread::get_id()<<std::endl;

//      for (;;)
//      {

//        std::cout<<"expires_after:"<<std::this_thread::get_id()<<std::endl;
//        timer_.expires_after(boost::asio::chrono::seconds(10));
//        std::size_t n = socket_.async_read_some(boost::asio::buffer(data), yield);
//        std::cout<<"received:"<<n<<" bytes"<<std::endl;

//        boost::asio::post(io_context, boost::bind(&session::otherWork, shared_from_this(), n));

//      }
//    }
//    catch (std::exception& e)
//    {
//      socket_.close();
//      timer_.cancel();
//    }
//  }

//  void timeout(boost::asio::yield_context yield)
//  {
//    while (socket_.is_open())
//    {

//      std::cout<<"timeout:"<<std::this_thread::get_id()<<std::endl;
//      boost::system::error_code ignored_ec;
//      timer_.async_wait(yield[ignored_ec]);
//      if (timer_.expiry() <= boost::asio::steady_timer::clock_type::now())
//      {
//          //std::cout<<"expiry:"<<timer_.expiry().time_since_epoch().count() / 1000000000<<std::endl;
//      }

//    }
//  }
//  void otherWork(std::size_t len)
//  {
//      std::cout<<"send:"<<len<<" bytes"<<std::endl;
//      //boost::asio::async_write(socket_, boost::asio::buffer(data, len), yield);
//      boost::asio::write(socket_, boost::asio::buffer(data, len));
//      std::this_thread::sleep_for(std::chrono::seconds(5));
//  }
  boost::asio::io_context::strand strand_;
  tcp::socket socket_;
  boost::asio::steady_timer timer_;
  char data[128];
};

void do_accept(boost::asio::io_context& io_context, unsigned short port/*,boost::asio::yield_context yield*/)
{
  tcp::acceptor acceptor(io_context, tcp::endpoint(tcp::v4(), port));
  boost::system::error_code ec;
  boost::fibers::future<boost::system::error_code> f;
  std::future<boost::system::error_code> fut;
  for (;;)
  {

    try
      {
          boost::shared_ptr<session> new_session(new session(io_context));

          f= acceptor.async_accept(new_session->socket(), boost::asio::fibers::use_future([&](const boost::system::error_code& ec){
               return ec;
          })) ;

          ec = f.get();
          if (!ec)
          {
              new_session->go(io_context);
          }

      }

    catch (std::system_error& ec)
    {
        if (ec.code().value() == boost::asio::error::no_descriptors)
        {

            continue;
        }
        else if (ec.code().value() == boost::asio::error::operation_aborted) //主动关闭结束
        {

            break;
        }

    }
  }
}

int main(int argc, char* argv[])
{
  try
  {
//    if (argc != 2)
//    {
//      std::cerr << "Usage: echo_server <port>\n";
//      return 1;
//    }
    std::cout<<"main:"<<std::this_thread::get_id()<<std::endl;
    boost::asio::io_context io_context;
    std::thread th([&]{boost::asio::io_context::work wk(io_context);io_context.run();});
    th.detach();
//    std::thread th1([&]{boost::asio::io_context::work wk(io_context);io_context.run();});
//    th1.detach();
//    boost::asio::spawn(io_context,
//        boost::bind(do_accept,
//          boost::ref(io_context), 8000, _1));

    boost::fibers::fiber([&]{do_accept(boost::ref(io_context), 8000);});
    //do_accept(boost::ref(io_context), 8000);
    getchar();

  }
  catch (std::exception& e)
  {
    std::cerr << "Exception: " << e.what() << "\n";
  }

  return 0;
}

acceptor.async_accept那句话在vs2017下老是报类型错误，能否帮我看一下，谢谢

**Comment by jxfwinter** [comment]
acceptor.async_accept<tcp, tcp::socket::executor_type>(new_session->socket(), boost::asio::fibers::use_future([&](const boost::system::error_code& ec){
           return ec;
      })) ;
